feat(cli): reload environment variables with `--watch` and `--env-file`

meetdhanani17 opened the pull request:
**fix(cli/watch): Reload environment variables with --watch and --env-file**

- Added functionality to watch the env file and reload its variables during hot reload execution.
- fixes issue [#29483](https://github.com/denoland/deno/issues/29483), [#29201](https://github.com/denoland/deno/issues/29201)

meetdhanani17 committed:
enhance environment file handling in CLI options

meetdhanani17 committed:
added lint fixes

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/denoland/deno?pullRequest=30197) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/denoland/deno?pullRequest=30197) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/denoland/deno?pullRequest=30197) it.</sub>

dsherret commented on cli/tools/run/mod.rs:
```diff
@@ -178,6 +181,12 @@ async fn run_with_watch(
+        load_env_variables_from_env_file(
+          cli_options.env_file_name(),
+          log_level,
+        );
```
I think you want to move this above creating the factory and use `flags.env_file` instead. That way the changed environment variables propagate into everything created by the factory.

meetdhanani17 replied:
Thank you for the suggestions, I'll make changes for it!

dsherret reviewed (commented):
Thanks! Would it be possible to add a test for this to ensure it works and continues to work in the future?

meetdhanani17 committed:
feat(cli): enhance environment file handling in CLI options

meetdhanani17 commented:
> I think this is a duplicate of #29555 ?

no it's different, you can take reference of the code changes

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

Signed-off-by: Meet Dhanani <124246235+meetdhanani17@users.noreply.github.com>

meetdhanani17 committed:
resolve conflicts of merge main

meetdhanani17 committed:
updated with main branch

meetdhanani17 committed:
resolve mismatch issues with main branch

meetdhanani17 committed:
fix linting issues for env_loader

meetdhanani17 committed:
fix lint issues for env_manager

meetdhanani17 committed:
added copyright message

meetdhanani17 committed:
fix test cases of load env for multiple files

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
optimize env variable loading functionality

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
fix lint issues for env_manager

meetdhanani17 committed:
fix lint issues

meetdhanani17 committed:
fix lint issue

meetdhanani17 commented:
Hello @dsherret dsherret, Could you please guild me with fixes the test cases?

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
optimize env load functionality according to existing functionality

meetdhanani17 committed:
fix formatting issue

meetdhanani17 committed:
remove unused import dotenvy::from_filename;

meetdhanani17 committed:
load env before init otel config

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

dsherret commented on cli/util/env_manager.rs:
```diff
@@ -0,0 +1,266 @@
+    log_level: Option<log::Level>,
+    inner: &mut EnvManagerInner,
+  ) {
+    let path_str = file_path.as_ref().to_string_lossy().to_string();
```
Recommend to change the function to receive a resolved `&Path`, then use `file_path.display()` in the eprintlns.

dsherret reviewed (dismissed):
Thanks! This PR seems to be on the right track.

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
change EnvManager to WatchEnvTracker for improved environment variable management
- Added WatchEnvTracker to track and manage environment variables loaded from .env files.
- Updated `resolve_flags_and_init` and `run_with_watch` functions to utilize the new tracker.
- Enhanced error handling for loading environment variables from files.

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

meetdhanani17 committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

dsherret committed:
Merge branch 'main' into fix/env_not_update_in_watch_mode

dsherret reviewed (approved):
LGTM. Thanks!
